### PR TITLE
SETI-401: Test some corner cases of abort-pipeline feature

### PR DESCRIPTION
**Pull request by janpriessnitz**
Test that abort-pipeline does not abort same change number but
from different project.

### CODE REVIEW

**Comment by janpriessnitz** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/107/ : SUCCESS in 11s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/127/ : FAILURE in 11m 04s

**Comment by janpriessnitz** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/108/ : SUCCESS in 8s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/128/ : SUCCESS in 11m 16s

File: zuul/scheduler.py [hunk 1]
```diff
@@ -1424,7 +1424,9 @@ def removeAllVersionsOfChange(self, change):
         self.log.debug("Removing versions of change %s in pipeline %s" %
                        (change, self.pipeline.name))
         for item in self.pipeline.getAllItems():
-            if item.change.number == change.number:
+            # Works for both Github PRs and Gerrit
+            if (item.change.number == change.number
+                and item.change.project == change.project):
```
Comment (hrubi) [inline zuul/scheduler.py]: You should delegate this to a new change method. For gerrit is enough to check for the `number` equality, while github needs to check also the `project` equality.

The only problem here is the naming, as there is already `Changeish.equals`, which checks also the equality of the "patchset".

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/113/ : SUCCESS in 18s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/137/ : SUCCESS in 13m 58s

File: tests/test_github.py [hunk 2]
```diff
@@ -574,3 +574,45 @@ def test_multiple_projects(self):
...
+        self.assertEqual(len(self.history), 4)
+        self.assertEqual(self.history[0].name, 'project-check1')
+        self.assertEqual(self.history[0].result, 'ABORTED')
+
+        run_job_names = ['project1-check1', 'project-gate', 'project1-check2']
+        self.assertIn(self.history[1].name, run_job_names)
+        self.assertIn(self.history[2].name, run_job_names)
+        self.assertIn(self.history[3].name, run_job_names)
+
+        self.assertEqual(A.statuses['gate']['state'], 'success')
```
Comment (hrubi) [inline tests/test_github.py]: Please assert the status of `A.statuses['check']['state']`, I wonder how does it look like.

File: tests/test_github.py [hunk 3]
```diff
@@ -574,3 +574,45 @@ def test_multiple_projects(self):
...
+        self.waitUntilSettled()
+
+        self.assertEqual(len(self.history), 4)
+        self.assertEqual(self.history[0].name, 'project-check1')
+        self.assertEqual(self.history[0].result, 'ABORTED')
+
+        run_job_names = ['project1-check1', 'project-gate', 'project1-check2']
+        self.assertIn(self.history[1].name, run_job_names)
+        self.assertIn(self.history[2].name, run_job_names)
+        self.assertIn(self.history[3].name, run_job_names)
```
Comment (hrubi) [inline tests/test_github.py]: I suppose you assert it this way because the order in `history` is not stable. What about:
```python
self.assertEqual(
```
I suppose you assert it this way because the order in `history` is not stable. Would this work for you?
```python
self.assertEqual(sorted(self.history.sorted), sorted(run_job_names))
```

File: tests/fixtures/layout-abort-pipelines.yaml [hunk 4]
```diff
@@ -39,3 +54,9 @@ projects:
         - project-check2
     gate:
       - project-gate
+  - name: org/project1
+    check:
+      - project1-check1:
+        - project1-check2
```
Comment (hrubi) [inline tests/fixtures/layout-abort-pipelines.yaml]: Is is necessary to have 2 jobs in check for org/project1 and org/project? Is it testing something additional than it would with 1 check job per project? If it's not, please remove it, it's a bit complicating the job counting in the test for me.

I mean.. in the test there is:
Open PR in A and PR in B:
check: 2 (A and B)
gate: 0

Add merge label to A:
check: 1 (B; check for A was aborted)
gate: 1 (A)

Suddenly after letting the jobs finish (this is what surprised me, but it's correct):
`self.assertEqual(len(self.history), 4)` - so this is supposedly: project-check1 (aborted), project-gate, project1-check1, project1-check2
project-check2 was not counted in self.history, because it did not start at all.

It would be a bit less complicated without the `project*-check2` jobs.

File: tests/test_github.py (hunk 2 above)
Comment (hrubi) [reply tests/test_github.py]: For this, you want to add to the layout check pipe:
```yaml
pipelines:
  - name: check
...
    start:
      github:
        status: true

```

The status remains in `pending`, which does not look desired.

File: tests/test_github.py (hunk 3 above)
Comment (janpriessnitz) [reply tests/test_github.py]: No, because self.history is a list of objects, but run_job_names is a list of strings. I'll try to clean it up in some other way.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/116/ : SUCCESS in 9s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/140/ : FAILURE in 10m 37s

**Comment by janpriessnitz** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/117/ : SUCCESS in 9s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/141/ : SUCCESS in 8m 40s

File: tests/fixtures/layout-abort-pipelines.yaml [hunk 5]
```diff
@@ -2,40 +2,55 @@
...
-  - name: project-check1
-  - name: project-check2
-  - name: project-gate
 
 projects:
   - name: org/project
     check:
-      - project-check1:
-        - project-check2
+      - project-check1
```
Comment (hrubi) [inline tests/fixtures/layout-abort-pipelines.yaml]: You even don't need to call it `check1` now, plain `check` would do. The same below.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/119/ : SUCCESS in 8s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/143/ : FAILURE in 9m 48s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/zuul-merge-compliance-check/23/ : SUCCESS in 12s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/144/ : FAILURE in 9m 42s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/120/ : SUCCESS in 12s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/145/ : FAILURE in 16m 06s

**Comment by janpriessnitz** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/121/ : SUCCESS in 10s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/146/ : SUCCESS in 7m 24s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/zuul-merge-compliance-check/24/ : SUCCESS in 11s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/147/ : SUCCESS in 8m 24s
